buildspec.yml

```diff
@@ -0,0 +1,33 @@
+version: 0.2
+
+phases:
+  build:
+    commands:
+      - git clone https://github.com/crungehottman/exodus-lambda.git
+      - cd exodus-lambda
+      - git checkout codebuild
+      - gpg --import .whitelist.gpg
+      - LAST_TAGGED=$(git rev-list --tags --max-count=1)
+      - git verify-tag $(git describe $LAST_TAGGED)
+      - git checkout $LAST_TAGGED
+      - pip install --require-hashes -r requirements.txt --target ./package
+      - pip install --no-deps --target ./package .
+      - cp ./configuration/exodus-lambda-deploy.yaml ./package
+      - >-
+        jq -n "{
```

Comment:
Could we find a way to have this template in a separate file in the repo rather than embedded into this YAML?

It would be neat if the template file could use YAML & Jinja for consistency with ansible.

Reply:
Actually, taking it further: I think everything here after the git verification and checkout steps should not be embedded directly in the buildspec.yml.  They could instead be hosted in a script in the repo (so that the buildspec would just do: verify & checkout desired revision, then run e.g. "scripts/build-package" or whatever).

The reason is: we didn't design anything to automatically deploy this buildspec.yml into AWS, did we?  Does that mean every time you need to change it, someone is expected to manually run some commands locally to deploy it?

If so, we should try to have the narrowest possible scope in the buildspec so the need for such changes would be minimized, because every change needs some special action beyond the usual release workflow, which is prone to human error (e.g. a developer accidentally deploying some locally modified version of the buildspec.yml rather than what's in git).

So, for example, if we need to change some detail of the config file template, or slightly change the "aws cloudformation ..." command which is run, ideally we can achieve that by the same workflow as used for changes to the lambdas themselves: merge to deployment branch and make a new signed tag.  No "deploy updated buildspec" step which could be forgotten or executed incorrectly.

Reply:
> Does that mean every time you need to change it, someone is expected to manually run some commands locally to deploy it?
No, changes to the buildspec would be applied when pushed to the deploy branch, like any other change to exodus-lambda source.

Reply:
So, just to be clear then: the repo is checked out more than once?  At least once to deploy the buildspec and then again while running the buildspec?

If buildspec changes checked into the repo are applied, then it's futile to try to add any kind of verification into the buildspec steps itself as the person who managed to inject their malicious commit can just strip out those verification steps from the buildspec.

Reply:
> [...] the repo is checked out more than once?

No, the pipeline receives a zip of the repo's files, in which the buildspec is included, whenever a commit is pushed to the deploy branch. It's this copy of the buildspec that's used to clone and checkout the full repo.

> [...] it's futile to try to add any kind of verification into the buildspec steps itself as the person who managed to inject their malicious commit can just strip out those verification steps from the buildspec.

Yeah...this was an oversight. I'm for "stripping out the GPG verification stuff here and instead shifting that into the scope of RHELDST-25."